azurelinuxagent/common/protocol/wire.py

```diff
@@ -862,6 +862,10 @@ def get_gafamily_manifest(self, vmagent_manifest, goal_state):
                     self.update_goal_state(forced=True)
                     goal_state = self.get_goal_state()
 
+                self._remove_stale_agent_manifest(
+                    vmagent_manifest.family,
+                    goal_state.incarnation)
+
                 local_file = MANIFEST_FILE_NAME.format(
                                 vmagent_manifest.family,
                                 goal_state.incarnation)
```

Comment:
It looks like it writes out the agent manifest file below but where does it prefer what it has on disk rather than what it just fetched over the network?

Should the `write_file` call below not just overwrite whatever was there before? How does doing an `rm` first change anything?

Reply:
The agents manifest file is computed, and written to disk in a different function.  The different function also does a little dance depending upon the file existing or not, which is why I have to delete it here.  The manifest file is written whenever a new goal state arrives.

Reply:
It looks like we can just skip the call to `filter_package_list` if we're never going to use the disk cache anyways. Just have `get_vmagent_pkgs` return `ga_manifest.pkg_list` instead.

Reply:
The cache file is written when the new goal state arrives because the agent manifest can change over time.  The idea is to snapshot the current value when the goal state arrives to ensure you only update the agent within a goal state.  

It also provides a historical record when we have to debug issues.